Assets/Scripts/BoardManager.cs

```diff
@@ -22,47 +44,168 @@ public class BoardManager : MonoBehaviour
+        columnsCount = Random.Range(columns.minimum, columns.maximum + 1);
+        rowsCount = Random.Range(rows.minimum, rows.maximum + 1);
+
+        numberOfRoom = Random.Range(numberOfRoomParts.minimum, numberOfRoomParts.maximum + 1) - 1;
+
+        boardHolder = new GameObject("Board").transform;
+
+        int x = 0;
+        int y = 0;
+        List<string> roomPosition = new List<string>();
+        List<Vector2> m = new List<Vector2>() { new Vector2(x, y) };
+        List<string> validPosition = new List<string>() { "top", "right", "bottom" };
+
+        //Calculate layout of other parts of the room
+        for (int i = 0; i < numberOfRoom; i++)
+        {
+            int randomIndex = Random.Range(0, 3);
+            if (validPosition[randomIndex] == "top")
+            {
+                roomPosition.Add("top");
+                y++;
+                m.Add(new Vector2(x, y));
+                validPosition = new List<string>() { "top", "right", "left" };
+            }
+            else if (validPosition[randomIndex] == "right")
+            {
+                roomPosition.Add("right");
+                x++;
+                m.Add(new Vector2(x, y));
+                validPosition = new List<string>() { "top", "right", "bottom" };
+            }
+            else if (validPosition[randomIndex] == "bottom")
+            {
+                roomPosition.Add("bottom");
+                y--;
+                m.Add(new Vector2(x, y));
+                validPosition = new List<string>() { "right", "bottom", "left" };
+            }
+            else if (validPosition[randomIndex] == "left")
+            {
+                roomPosition.Add("left");
+                x--;
+                m.Add(new Vector2(x, y));
+                validPosition = new List<string>() { "top", "bottom", "left" };
+            }
+        }
+
+        //Generate first room part
+        GenerateRoomPart(x, y, m, 0);
 
-        for (int x = 1; x < columns - 1; x++)
+        //Generate other room parts
+        int index = 1;
+        foreach (var pos in roomPosition)
         {
-            for (int y = 1; y < rows - 1; y++)
+            if (pos == "top")
+            {
+                y += rowsCount;
+                GenerateRoomPart(x, y, m, index); //top
+            }
+            else if (pos == "right")
+            {
+                x += columnsCount;
+                GenerateRoomPart(x, y, m, index); //right
+            }
+            else if (pos == "bottom")
+            {
+                y -= rowsCount;
+                GenerateRoomPart(x, y, m, index); //bottom
+            }
+            else if (pos == "left")
             {
-                gridPositions.Add(new Vector3(x, y, 0f));
+                x -= columnsCount;
+                GenerateRoomPart(x, y, m, index); //left
             }
+            index++;
         }
     }
 
-    void BoardSetup()
+    void GenerateRoomPart(int xNumber, int yNumber, List<Vector2> m, int roomPos)
```

Comment:
I suppose xNumber and yNumber are the x/y starting position of the room ?
If so maybe rename them to x/yStart or any more explicit name.